Fix inode eviction and sync writeback deadlock on Linux

tuxoko opened the pull request:
If inode eviction and sync writeback happen on the same inode at the same time, inode eviction will set I_FREEING and wait for sync writeback, and sync writeback may eventually calls zfs_get_data and loop in zfs_zget forever because igrab cannot succeed with I_FREEING, thus causing deadlock.

To fix this, we take a hold on znode in itx. This make sure the inode won't get evicted until sync writeback is done.

Signed-off-by: Chunwei Chen <david.chen@nutanix.com>
Fixes #7964
Fixes #9430

<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

### Description
<!--- Describe your changes in detail -->

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

tuxoko commented on module/zfs/zfs_vnops.c:
```diff
@@ -1149,10 +1149,21 @@ zfs_get_data(void *arg, uint64_t gen, lr_write_t *lr, char *buf,
 	ASSERT3U(size, !=, 0);
 
+	error = zfs_zget_impl(zfsvfs, object, &zp, B_TRUE);
+#if defined(__linux__)
```
I added ifdef because I'm not sure if it's possible to return EAGAIN in freebsd or not.
But I guess looking at the caller, it doesn't really matter if we return EAGAIN or EIO.

snajpa replied:
perhaps it would be more clean to enhance `zil_lwb_commit`, which is the user of `zfs_get_data`, to recognize `EAGAIN`, I think

tonyhutter reviewed (approved):
I'm not an expert in the znode code, but I don't see any surface level issues.

snajpa commented on module/os/linux/zfs/zfs_znode_os.c:
```diff
@@ -1109,8 +1117,18 @@ zfs_zget(zfsvfs_t *zfsvfs, uint64_t obj_num, znode_t **zpp)
+				 * path, i.e. zfs_get_data, if inode is being
+				 * evicted and I_SYNC is also set.
+				 */
+				if (check_sync && (ZTOI(zp)->i_state & I_SYNC))
```
maybe we can move this before the `igrab`, take the `i_lock` and check for `I_FREEING` and `I_SYNC` and if they are both set, bail out of the original `zfs_zget` without the need for `zfs_zget_impl`

tuxoko replied:
The only place that needs bail out is if you are in writeback context, either you are the one setting I_SYNC or whoever set it is blocked by you.

If you do bailout for ever caller than you will get error randomly in all sorts of places for no good reason.

snajpa replied:
`I_SYNC` is set in two kernel functions, `writeback_sb_inodes`, `writeback_single_inode`. So if you check for `I_SYNC` and `I_FREEING` and you use `i_lock` correctly as you're supposed to, you get more clean and also reliable solution. How is what am saying wrong _specifically_ and why? Which tons of errors?

tuxoko replied:
@snajpa 
I_FREEING and I_SYNC is set on the inode. Every one trying to do zfs_zget will see them. So if you don't have any restriction on caller then every one will bailout and got error as long as eviction and writeback is happening at the same time.

snajpa replied:
well.. I tried playing around with it and it seems like it would be an interesting project to get to the bottom of maybe few other lurking bugs, forcing one to go around and look at how errors from `zfs_zget` are handled all over the codebase - I think there's a thing in direct IO I'll try to tackle if I find time; but that's not relevant to resolving the deadlock, so, yeah, no easy way other than the `_impl` split :)

tonyhutter commented:
@snajpa would you mind taking another look at this one when you get a chance?

snajpa commented:
> @snajpa would you mind taking another look at this one when you get a chance?

@tonyhutter

I mean I don't like it, because now evicting inodes with dirty data will cause txg closing, as fallback of not being able to do `zfs_get_data` to generate the zil records for inodes open with `O_SYNC` at the very least -- in `zil_commit_impl` there's is `txg_wait_synced` for the failure case, if I understand the patch and those codepaths correctly. Great learning opportunity for me if I don't :D

I think we should do something specifically in case the inode is evicting and syncing both, when `zfs_zget_impl` returns `EAGAIN`.

@tuxoko how about this:

If there's `EAGAIN` and a flag is given to `zfs_zget_impl`, I think we could instead exit `zfs_zget` holding two locks - the `zfs_znode_hold`, but also `z_lock` is important here: `z_unlinked` should only be modified and checked under `z_lock` anyway - then the `znode` could be worked with for `zfs_get_data` purposes (and unlock and rele at the end)

or please demonstrate if I'm wrong in my expectation that this will cause txg flushing even in cases of memory cgroup triggered evictions, how am I wrong specifically - we can't be closing txgs on evict (in my very strongly held opinion, because I do have problems with closing txgs faster than 2-3 seconds on our setups as we rely on the writeback mechanisms in ZFS, also want them to reduce rewrites)

and also please consider that `i_state` is supposed to be accessed under `i_lock` spinlock, it might not matter specifically in this case but again in my opinion it's cheap to take it in the uncontended case so, what's the harm in doing it consistently with rest of Linux

tuxoko commented:
@snajpa 
I_SYNC writeback flag is for dirtied mmap pages. It's a separate thing from O_SYNC.
If you don't do mmap, then I don't think you will be affected by the deadlock or this change.

snajpa commented:
Lots of software do mmap, we in particular do everything, because we run containers in containers on big boxes and OpenZFS so far has been the most performant option for our use-case in practice, even though microbenchmarks on otherwise idle boxes might not agree :) If this ends up triggering txg flushes on in-container inode evicts, I'm going to have to modify it in some way (if as I outlined works then that way)

snajpa commented:
But then of course currently, due to the very bug you're trying to solve, we have to run with `sync=disabled`... so, you do have my grattitude for uncovering what's behind the deadlock, but please let's solve it better

tuxoko commented:
@snajpa 
If your performance is affected by the patch, then wouldn't you have already affected by the deadlock as well?

tuxoko commented:
Just to clarify, I don't disagree trying to gracefully get the znode and let zil commit continues normally would be better.
But either we have to make sure we are in writeback context or if we can somehow prevent inode being evicted under us.
Given this deadlock is not very common, I doubt this will have that much performance impact.
Though I'm welcome to be proven otherwise.

snajpa commented:
Depends, the only the first case of this happening currently deadlocks the machine so that's the one we know about :)

but if this happens continually in multiple cgroups in a loop and there isn't a deadlock, then we will still have to run with `sync=disabled` until a solution for that is in place, because the result would be a txg sync loop, not very keen on seeing those, loadavg goes up, meaning useful workload suffers by a few misbehaving workloads (out of my admin jurisdiction even as we're sort of a timesharing coop, members manage the first container level on the inside ;))

behlendorf commented:
> either we have to make sure we are in writeback context or if we can somehow prevent inode being evicted under us.

We could prevent the inode from being evicted by registering our own `drop_inode()` function.  The tricky bit is we'd want a cheap way to check if there are outstanding itx entries for the file which `zil_commit()` needs to handle.  Alternately, we could hold an inode reference until those same itx entries are processed.  That feels a bit nicer and would prevent the `final_iput()` from occurring and evicting the inode until we're really done with it.

snajpa commented:
> Alternately, we could hold an inode reference until those same itx entries are processed

Best approach in my opinion from a point of view of a running system under load, but that's going to affect a lot of (common) code, isn't it?

What is the rationale for merging inodes with znodes in OpenZFS, btw? Originally znode had a pointer to vnode IIRC; that could help here, `zfs_get_data` doesn't need the `ip`, so I'm wondering if maybe that wouldn't also solve the issue?

behlendorf commented:
> Best approach in my opinion from a point of view of a running system under load, but that's going to affect a lot of (common) code, isn't it?

I'm not sure how invasive it would be without trying it out.  We already have common `zhold` and `zrele` wrappers which could be used for this.

> What is the rationale for merging inodes with znodes in OpenZFS, btw?

I wanted to have a single source of truth and there's a lot of redundancy between the inode and znode fields.  Plus there's really isn't any direct equivalent for vnodes on Linux.

tuxoko commented:
> > either we have to make sure we are in writeback context or if we can somehow prevent inode being evicted under us.
> 
> We could prevent the inode from being evicted by registering our own `drop_inode()` function.

I'm not sure if that would work in general, drop_inode returning false would just put it back to lru, which could still be evict anytime.

> Alternately, we could hold an inode reference until those same itx entries are processed. That feels a bit nicer and would prevent the final_iput() from occurring and evicting the inode until we're really done with it.

By the time kernel writeback code calls into zfs, iput_final may have already happened though.

robn commented:
Sorry, I missed this ping.

It's a little difficult to follow exactly what's happening from the discussion, so I'll try to summarise what I think the discussion is about. Let me know if I've got this wrong.

The original issue is that inode eviction triggers a sync write, so `zil_commit()` -> `zfs_get_data()`, which loops forever trying to get reference on the inode through `zfs_zget()` -> `igrab()`, right? If that's so, the simplest resolution is for `zfs_get_data()` to return `EIO`. `zil_lwb_commit()` will fall back to `txg_wait_synced_flags()`, which will complete, and that's that.

The rest seems to be discussion how to avoid a full txg sync (and so preventing a stall) just because the kernel is evicting an inode, correct? That's a separate discussion, I think, because I have a few followup questions (not least: if the kernel has said "sync this", why do we think its ok to ignore that?). But mostly, I don't think its for this PR: I will basically always argue for correctness over performance (but I will shut up after saying my piece).

davidchenntnx committed:
Fix inode eviction and sync writeback deadlock on Linux

If inode eviction and sync writeback happen on the same inode at the
same time, inode eviction will set I_FREEING and wait for sync
writeback, and sync writeback may eventually calls zfs_get_data and loop
in zfs_zget forever because igrab cannot succeed with I_FREEING, thus
causing deadlock.

To fix this, we take a hold on znode in itx. This make sure the inode
won't get evicted until sync writeback is done.

Signed-off-by: Chunwei Chen <david.chen@nutanix.com>
Fixes #7964
Fixes #9430

tuxoko commented:
Update:
Rework the fix by holding znode in itx as suggested by @behlendorf .

jumbi77 commented:
Can i politely ping @behlendorf to give feedback?

alek-p commented on module/zfs/zil.c:
need to add:
```
itx->itx_znode = NULL;
itx->itx_zrele = NULL;
```

behlendorf commented on module/os/linux/zfs/zfs_vfsops.c:
```diff
@@ -1349,6 +1349,28 @@ zfsvfs_teardown(zfsvfs_t *zfsvfs, boolean_t unmounting)
+
 	/*
 	 * If someone has not already unmounted this file system,
 	 * drain the zrele_taskq to ensure all active references to the
```
Moving the `zil_close()` up seems okay, and the easiest way to handle this.

behlendorf reviewed (approved):
@tuxoko sorry I lost track of this (again).  Your updated version which holds the znodes makes sense to me.  Let's run it by @pcd1193182 though, he's recently been working in the area and may have some thoughts.

behlendorf commented:
There may be an issue on the FreeBSD side, all of the FreeBSD builders failed somewhat mysteriously.  I've resubmitted those builds to see if it's easily reproducible.

alek-p commented:
I guess `zhold()` was barely used on FreeBSD before this patch. It appears the problem is that on FreeBSD`zhold()` and `VN_HOLD()` use different ref counts for the same vnode:
```
include/os/freebsd/zfs/sys/zfs_znode_impl.h lines 106-107

  #define zhold(zp)   vhold(ZTOV((zp)))
  #define zrele(zp)   vrele(ZTOV((zp)))
```
`vhold()` uses `v_holdcnt`
`vrele()`) uses `v_usecount`
```
include/os/freebsd/spl/sys/vnode.h lines 98-99

  #define VN_HOLD(v)  vref(v)
  #define VN_RELE(v)  vrele(v)
```
while `vref()` (and `vrele()`) use `v_usecount`.

Since `vref` is what pair with `vrele` I'm betting if you include
```
diff --git include/os/freebsd/zfs/sys/zfs_znode_impl.h include/os/freebsd/zfs/sys/zfs_znode_impl.h
index 5ead67053..83e4dbdac 100644
--- include/os/freebsd/zfs/sys/zfs_znode_impl.h
+++ include/os/freebsd/zfs/sys/zfs_znode_impl.h
@@ -103,7 +103,12 @@ typedef struct zfs_soft_state {
 #define        VTOZ(VP)        ((struct znode *)(VP)->v_data)
 #define        VTOZ_SMR(VP)    ((znode_t *)vn_load_v_data_smr(VP))
 #define        ITOZ(VP)        ((struct znode *)(VP)->v_data)
-#define        zhold(zp)       vhold(ZTOV((zp)))
+#define        zhold(zp)       vref(ZTOV((zp)))
 #define        zrele(zp)       vrele(ZTOV((zp)))

 #define        ZTOZSB(zp) ((zp)->z_zfsvfs)
```
the panics go away.

amotin commented:
@alek-p I don't know much about VFS, but according to the man page, the opposite to `vhold()` would be a `vdrop()` that ZFS does not mention. So considering ZFS never used `zhold()` on FreeBSD before, and considering that it helps, I tend to think that your patch might be right.

davidchenntnx committed:
Freebsd use vref for zhold

tuxoko commented:
Update patch for freebsd zhold

alek-p commented:
the new run revealed a FreeBSD teardown ordering issue. Linux does teardown/`zil_close()` before the inode eviction but FreeBSD does not. Right now `zfs_unmout()` in FreeBSD does `vflush()` then `zil_close()` (called from `zfsvfs_teardown()`) meaning usecount isn't necessarily at zero when we `vflush()`.

Unfortionately,  I don't have a FreeBSD based zfs-tests setup where I can verify this but I think the following changes are in order:
```
diff --git module/os/freebsd/zfs/zfs_vfsops.c module/os/freebsd/zfs/zfs_vfsops.c
index fc54cef19..50b3e7572 100644
--- module/os/freebsd/zfs/zfs_vfsops.c
+++ module/os/freebsd/zfs/zfs_vfsops.c
@@ -1790,23 +1790,41 @@ zfs_umount(vfs_t *vfsp, int fflag)
                        return (ret);
        }

+       ZFS_TEARDOWN_ENTER_WRITE(zfsvfs, FTAG);
+       if (zfsvfs->z_log != NULL) {
+               zil_close(zfsvfs->z_log);
+               zfsvfs->z_log = NULL;
+       }
        if (fflag & MS_FORCE) {
                /*
                 * Mark file system as unmounted before calling
                 * vflush(FORCECLOSE). This way we ensure no future vnops
                 * will be called and risk operating on DOOMED vnodes.
                 */
-               ZFS_TEARDOWN_ENTER_WRITE(zfsvfs, FTAG);
                zfsvfs->z_unmounted = B_TRUE;
-               ZFS_TEARDOWN_EXIT(zfsvfs, FTAG);
+       }
+       ZFS_TEARDOWN_EXIT(zfsvfs, FTAG);
+
+       if (zfsvfs->z_os != NULL) {
+               taskq_wait_outstanding(dsl_pool_zrele_taskq(
+                   dmu_objset_pool(zfsvfs->z_os)), 0);
        }

        /*
         * Flush all the files.
         */
        ret = vflush(vfsp, 0, (fflag & MS_FORCE) ? FORCECLOSE : 0, td);
-       if (ret != 0)
+       if (ret != 0) {
+               ZFS_TEARDOWN_ENTER_WRITE(zfsvfs, FTAG);
+               ASSERT0P(zfsvfs->z_log);
+               if (zfsvfs->z_os != NULL) {
+                       zfsvfs->z_log = zil_open(zfsvfs->z_os, zfs_get_data,
+                           &zfsvfs->z_kstat.dk_zil_sums);
+               }
+               ZFS_TEARDOWN_EXIT(zfsvfs, FTAG);
                return (ret);
+       }
        while (taskqueue_cancel(zfsvfs_taskq->tq_queue,
            &zfsvfs->z_unlinked_drain_task, NULL) != 0)
                taskqueue_drain(zfsvfs_taskq->tq_queue,
```
This patch for FreeBSD's `zfs_unmount()` does 3 things:
1) Call `zil_close()` *before* `vflush()` to queue up `vrele()`s
2) Drain all async `vrele()` to actually drop the usecount
3) If `vflush()` failed - re-open the ZIL

davidchenntnx committed:
FreeBSD do zil_close before vflush during unmount

behlendorf commented:
I resubmitted the failed builds, but I think with Alek's last batch of suggested FreeBSD changes we should be in good shape here.  Two of previous FreeBSD builds failed when started up the test environment, and the third ran to completion but had a single unrelated test failure.

@tuxoko when you get a chance could you rebase this and squash the commits.
